File: tutorials/30-days-of-code/30-binary-trees.py
# Tutorials > 30 Days of Code > Day 23: BST Level-Order Traversal
# Implement a breadth-first search!
#
# https://www.hackerrank.com/challenges/30-binary-trees/problem
#

import sys

class Node:
    def __init__(self,data):
        self.right=self.left=None
        self.data = data
class Solution:
    def insert(self,root,data):
        if root==None:
            return Node(data)
        else:
            if data<=root.data:
                cur=self.insert(root.left,data)
                root.left=cur
            else:
                cur=self.insert(root.right,data)
                root.right=cur
        return root
# (skeliton_head) ----------------------------------------------------------------------

    def levelOrder(self,root):
        queue = []
        result = []

        queue.append(root)
        while queue:
            n = queue.pop(0)
            if n:
                result.append(n.data)
                queue.append(n.left)
                queue.append(n.right)

        print(*result)

# (skeliton_tail) ----------------------------------------------------------------------
T=int(input())
myTree=Solution()
root=None
for i in range(T):
    data=int(input())
    root=myTree.insert(root,data)
myTree.levelOrder(root)

File: tutorials/30-days-of-code/30-binary-trees.cpp
// Tutorials > 30 Days of Code > Day 23: BST Level-Order Traversal
// Implement a breadth-first search!
//
// https://www.hackerrank.com/challenges/30-binary-trees/problem
//

#include <iostream>
#include <cstddef>
#include <queue>
#include <string>
#include <cstdlib>

using namespace std;
class Node{
    public:
        int data;
        Node *left,*right;
        Node(int d){
            data=d;
            left=right=NULL;
        }
};
class Solution{
    public:
  		Node* insert(Node* root, int data){
            if(root==NULL){
                return new Node(data);
            }
            else{
                Node* cur;
                if(data<=root->data){
                    cur=insert(root->left,data);
                    root->left=cur;
                }
                else{
                   cur=insert(root->right,data);
                   root->right=cur;
                 }
           return root;
           }
        }
// (skeliton_head) ----------------------------------------------------------------------

    void levelOrder(Node * root)
    {
        queue<Node *>   q;

        q.push(root);
        while (! q.empty())
        {
            auto n = q.front();
            q.pop();

            if (n)
            {
                cout << n->data << " ";

                q.push(n->left);
                q.push(n->right);
            }
        }

        cout << endl;
    }

// (skeliton_tail) ----------------------------------------------------------------------
};//End of Solution
int main(){
    Solution myTree;
    Node* root=NULL;
    int T,data;
    cin>>T;
    while(T-->0){
        cin>>data;
        root= myTree.insert(root,data);
    }
    myTree.levelOrder(root);
    return 0;
}
